pkg/types/azure/validation/machinepool.go

```diff
@@ -221,6 +235,91 @@ func validateOSImage(p *azure.MachinePool, fldPath *field.Path) field.ErrorList
 	return allErrs
 }
 
+// getDataDisksSecurityProfileType extracts the security encryption type from data disks
+// to ensure consistency with OS disk security settings.
+//
+// Security profile requirements (from CAPZ types.go):
+// - When set to VMGuestStateOnly: VirtualizedTrustedPlatformModule should be Enabled
+// - When set to DiskWithVMGuestState: EncryptionAtHost should be disabled,
+//   SecureBoot and VirtualizedTrustedPlatformModule should be Enabled
+// - DiskWithVMGuestState can only be used with Confidential VMs
+//
+// Note: All data disks should have the same security encryption type to avoid
+// configuration conflicts and ensure consistent security posture.
+
+func getDataDisksSecurityProfileType(p *azure.MachinePool) capz.SecurityEncryptionType {
+	for _, dataDisk := range p.DataDisks {
+		if dataDisk.ManagedDisk != nil && dataDisk.ManagedDisk.SecurityProfile != nil {
+			switch dataDisk.ManagedDisk.SecurityProfile.SecurityEncryptionType {
+			case capz.SecurityEncryptionTypeDiskWithVMGuestState:
+				return capz.SecurityEncryptionTypeDiskWithVMGuestState
+			case capz.SecurityEncryptionTypeVMGuestStateOnly:
+				return capz.SecurityEncryptionTypeVMGuestStateOnly
+
+			default:
+				continue
```

Comment:
:question: IIUC, we want to make sure to only get the _valid_ security encryption type right? Should we add a validation for that?